File: sessao-5/tutorial-09-engenharia-de-prompt/exemplos/preco_revisado.py
"""
VERSÃO REVISADA — cálculo de preço com descontos após revisão de código gerado por IA
Referência: Clean Code, Cap. 2–3; engenharia de contexto em prompts de código

Problemas corrigidos em relação à versão gerada:
  - Nomes descritivos em português para todos os identificadores
  - Parâmetros soltos substituídos por @dataclass ItemPedido
  - Números mágicos extraídos como constantes nomeadas
  - Regras de desconto separadas em funções próprias
  - Acumulação indevida de descontos eliminada (aplica apenas o maior)
  - Validação com exceção descritiva em vez de comportamento silencioso

Nota didática: mesmo a partir de um prompt estruturado, a IA gerou a
comparação de descontos inline na função principal — foi preciso extrair
`_selecionar_maior_desconto` manualmente para manter a responsabilidade única.

Execute: python3 sessao-5/tutorial-09-engenharia-de-prompt/exemplos/preco_revisado.py
"""

from __future__ import annotations

from dataclasses import dataclass

DESCONTO_VOLUME_PCT = 0.10       # desconto por quantidade elevada
QUANTIDADE_MINIMA_VOLUME = 5     # quantidade mínima para desconto por volume
DESCONTO_PREMIUM_PCT = 0.15      # desconto para itens da categoria premium
SEM_DESCONTO = 0.0               # sentinela para ausência de desconto


@dataclass
class ItemPedido:
    descricao: str
    preco_unitario: float
    quantidade: int
    categoria: str = "padrao"


def _desconto_por_volume(item: ItemPedido) -> float:
    """Retorna o percentual de desconto por volume, ou SEM_DESCONTO."""
    if item.quantidade >= QUANTIDADE_MINIMA_VOLUME:
        return DESCONTO_VOLUME_PCT
    return SEM_DESCONTO


def _desconto_por_categoria(item: ItemPedido) -> float:
    """Retorna o percentual de desconto por categoria, ou SEM_DESCONTO."""
    if item.categoria == "premium":
        return DESCONTO_PREMIUM_PCT
    return SEM_DESCONTO


def _selecionar_maior_desconto(item: ItemPedido) -> float:
    """Retorna o maior desconto aplicável ao item (descontos não se acumulam)."""
    return max(_desconto_por_volume(item), _desconto_por_categoria(item))


def calcular_preco_final(item: ItemPedido) -> float:
    """
    Calcula o preço final do item aplicando o maior desconto elegível.

    Lança ValueError se preco_unitario for menor ou igual a zero.
    """
    if item.preco_unitario <= 0:
        raise ValueError(
            f"Preço unitário inválido para '{item.descricao}': {item.preco_unitario}. "
            "O valor deve ser maior que zero."
        )

    desconto = _selecionar_maior_desconto(item)
    preco_com_desconto = item.preco_unitario * (1 - desconto) * item.quantidade
    return round(preco_com_desconto, 2)


# ─── Execução de demonstração ─────────────────────────────────────────────────

if __name__ == "__main__":
    item_simples = ItemPedido(
        descricao="Camiseta Básica",
        preco_unitario=50.0,
        quantidade=3,
    )
    print("Preço item simples (sem desconto):", calcular_preco_final(item_simples))

    item_volume = ItemPedido(
        descricao="Camiseta Básica",
        preco_unitario=50.0,
        quantidade=6,
    )
    print("Preço com desconto por volume:", calcular_preco_final(item_volume))

    item_premium = ItemPedido(
        descricao="Tênis Premium",
        preco_unitario=200.0,
        quantidade=2,
        categoria="premium",
    )
    print("Preço com desconto premium:", calcular_preco_final(item_premium))

    # volume + premium: aplica apenas o maior (premium = 15% > volume = 10%)
    item_premium_volume = ItemPedido(
        descricao="Tênis Premium",
        preco_unitario=200.0,
        quantidade=6,
        categoria="premium",
    )
    print("Preço premium com volume (aplica apenas o maior):", calcular_preco_final(item_premium_volume))

    # preço inválido — deve lançar exceção
    try:
        item_invalido = ItemPedido(descricao="Produto Inválido", preco_unitario=0.0, quantidade=1)
        calcular_preco_final(item_invalido)
    except ValueError as erro:
        print(f"\nValidação: {erro}")

File: sessao-5/tutorial-09-engenharia-de-prompt/exemplos/preco_revisado.ts
/**
 * Revisão corrigida — Cálculo de Preço com Descontos
 * Referência: Tutorial 09 — Engenharia de contexto e prompt para gerar código
 * Execute: npx ts-node preco_revisado.ts
 *
 * Correção aplicada em relação a preco_gerado.ts:
 *   - Regra de negócio restaurada: aplica apenas o MAIOR desconto disponível,
 *     não a composição multiplicativa dos dois.
 *   - Demo idêntico ao gerado — resultado agora correto.
 */

// ─── Constantes de domínio ────────────────────────────────────────────────────

const DESCONTO_VOLUME_MEDIO      = 0.05;  // 5 % para 10–49 unidades
const DESCONTO_VOLUME_ALTO       = 0.10;  // 10 % para 50+ unidades
const DESCONTO_CATEGORIA_PREMIUM = 0.15;  // 15 % para categoria "premium"

const LIMIAR_VOLUME_MEDIO = 10;
const LIMIAR_VOLUME_ALTO  = 50;

// ─── Entidade ─────────────────────────────────────────────────────────────────

interface ItemPedido {
  produtoId:  string;
  categoria:  string;
  precoUnit:  number;
  quantidade: number;
}

// ─── Cálculo de preço ─────────────────────────────────────────────────────────

function calcularDescontoVolume(quantidade: number): number {
  if (quantidade >= LIMIAR_VOLUME_ALTO)  return DESCONTO_VOLUME_ALTO;
  if (quantidade >= LIMIAR_VOLUME_MEDIO) return DESCONTO_VOLUME_MEDIO;
  return 0.0;
}

function calcularDescontoCategoria(categoria: string): number {
  if (categoria === "premium") return DESCONTO_CATEGORIA_PREMIUM;
  return 0.0;
}

function calcularPrecoFinal(item: ItemPedido): number {
  /**
   * Calcula o preço final do item aplicando os descontos disponíveis.
   *
   * Regra de negócio: vale o MAIOR desconto — apenas um é aplicado.
   * Se o item for premium (15%) e tiver volume alto (10%), aplica 15%.
   */
  const subtotal = item.precoUnit * item.quantidade;

  const descontoVolume    = calcularDescontoVolume(item.quantidade);
  const descontoCategoria = calcularDescontoCategoria(item.categoria);

  // Regra correta: seleciona o MAIOR desconto disponível.
  const melhorDesconto = Math.max(descontoVolume, descontoCategoria);
  return subtotal * (1.0 - melhorDesconto);
}

function formatarResultado(item: ItemPedido, precoFinal: number): string {
  const subtotal = item.precoUnit * item.quantidade;
  const economia = subtotal - precoFinal;
  return (
    `  ${item.produtoId} (${item.categoria}): ` +
    `subtotal=R$${subtotal.toFixed(2)}  final=R$${precoFinal.toFixed(2)}  ` +
    `economia=R$${economia.toFixed(2)}`
  );
}

// ─── Execução de demonstração ─────────────────────────────────────────────────

const itens: ItemPedido[] = [
  { produtoId: "P001", categoria: "eletronicos", precoUnit: 100.00, quantidade:  5 },  // sem desconto
  { produtoId: "P002", categoria: "eletronicos", precoUnit: 100.00, quantidade: 20 },  // só volume médio (5%)
  { produtoId: "P003", categoria: "premium",     precoUnit: 200.00, quantidade:  3 },  // só categoria (15%)
  { produtoId: "P004", categoria: "premium",     precoUnit: 200.00, quantidade: 60 },  // maior = categoria (15%)
];

console.log("=== Cálculo de Preço (revisado — regra de negócio correta) ===\n");

for (const item of itens) {
  const preco = calcularPrecoFinal(item);
  console.log(formatarResultado(item, preco));
}

console.log();
console.log("Caso crítico — P004 (premium, 60 unidades):");
const itemCritico = itens[3];
const subtotalCritico = itemCritico.precoUnit * itemCritico.quantidade;
const precoCorreto = calcularPrecoFinal(itemCritico);
const precoEsperado = subtotalCritico * (1.0 - DESCONTO_CATEGORIA_PREMIUM);
console.log(`  subtotal             = R$${subtotalCritico.toFixed(2)}`);
console.log(`  desconto maior (15%) = R$${(subtotalCritico * DESCONTO_CATEGORIA_PREMIUM).toFixed(2)}`);
console.log(`  preço esperado       = R$${precoEsperado.toFixed(2)}`);
console.log(`  preço calculado      = R$${precoCorreto.toFixed(2)}  <- CORRETO`);
